File: src/main/scala/universalelectricity/core/grid/NodeGrid.scala
package universalelectricity.core.grid

import scala.collection.convert.wrapAll._
import scala.reflect.ClassTag

/**
 * A grid that consists of specific INodes
 *
 * @tparam N - The type of the node
 */
abstract class NodeGrid[N <: Node[_] : ClassTag] extends Grid[N]
{
  protected override def reconstructNode(node: N)
  {
    node.recache()
    node.setGrid(this)

    val connections = node.connections

    connections.keySet().foreach(
      connectedNode =>
      {
        if (isValidNode(connectedNode))
        {
          val con = connectedNode.asInstanceOf[N]

          if (con.getGrid != this)
          {
            con.getGrid.getNodes().clear()
            add(con)
            reconstructNode(con)
          }
        }
      })
  }

  override def deconstruct
  {
    getNodes().foreach(
      node =>
      {
        node.setGrid(null)
        node.reconstruct()
      })

    getNodes().clear()
  }
}

File: src/main/scala/universalelectricity/core/grid/NodeGrid.java
package universalelectricity.core.grid;

import net.minecraftforge.common.util.ForgeDirection;

import java.util.Map;

/**
 * Created by robert on 8/13/2014.
 */
public class NodeGrid<N extends Node> extends Grid<N>
{
    public NodeGrid(Class node)
    {
        super(node);
    }

    @Override
    protected void reconstructNode(N node)
    {
        node.recache();
        node.setGrid(this);

        Map<N, ForgeDirection> connections = node.connections();

        for(Map.Entry<N, ForgeDirection> entry : connections.entrySet())
        {
            if (isValidNode(entry.getKey()))
            {
                N con = (N)entry.getKey();

                if (con.getGrid() != this)
                {
                    //TODO redo merge grid logic
                    con.getGrid().getNodes().clear();
                    add(con);
                    reconstructNode(con);
                }
            }
        }
    }

    @Override
    public void deconstruct()
    {
        for(N node : getNodes())
        {
            node.setGrid(null);
            node.reconstruct();
        }
        super.deconstruct();
    }
}
